File: Knapsack/knapsack.cpp
#include <iostream>
#include <algorithm>

const int NUM_ITEMS = 10;
const int CAPACITY = 10;

int knapsack(int weights[NUM_ITEMS], int values[NUM_ITEMS], int n, int capacity, int memo[NUM_ITEMS + 1][CAPACITY + 1]){
    int result;
    if (memo[n][capacity] != -1){
        return memo[n][capacity];
    }

    if(n == 0 || capacity == 0){
        result = 0;
    } else if (weights[n] > capacity){
        result = knapsack(weights, values, n-1, capacity, memo);
    }else{

        int not_added = knapsack(weights, values, n-1, capacity, memo);
        int added = values[n] + knapsack(weights, values, n-1, capacity - weights[n], memo);
        result = std::max(not_added, added);
    }
    memo[n][capacity] = result;
    return result;
}


int main() {
    int v[NUM_ITEMS] = {2,4,6,2,7,9,3,2,5,10};
    int w[NUM_ITEMS] = {1,2,3,4,5,6,7,8,9,10};
    int memo[NUM_ITEMS + 1][CAPACITY + 1] = {{-1}};

    std::cout << "Max value: " << knapsack(w, v, NUM_ITEMS, CAPACITY, memo) << std::endl;
}

File: Knapsack/knapsack.rs
extern crate rand;
use std::time::Instant;
use std::cmp;
use rand::Rng;

fn main() {
    let num_items: i32 = 1000;
    let w = make_rand_vec(num_items);
    let v = make_rand_vec(num_items);
    let n = w.len() - 1;
    let c = num_items;
    let mut memo = vec![vec![-1; w.len() + 1]; (c + 1) as usize];
    let time = Instant::now();
    let max= knapsack(&w, &v, n, c as i32, &mut memo);
    let elapsed = time.elapsed().as_micros() as f64 / 1000.0;
    println!("Max value = {}", max);
    print!("Took {}ms", elapsed);
}

fn knapsack(weights: &Vec<i32>, values: &Vec<i32>, n: usize, capacity: i32, memo: &mut Vec<Vec<i32>>) -> i32 {
    let result: i32;
    if memo[n][capacity as usize] != -1 {
        return memo[n][capacity as usize] as i32;
    }

    if n == 0 || capacity == 0 {
        result = 0;
    }
    else if weights[n] > capacity {
        result = knapsack(weights, values, n-1, capacity, memo);
    }
    else {
        let not_added = knapsack(weights, values, n-1, capacity, memo);
        let added = values[n] + knapsack(weights, values,n-1, capacity - weights[n], memo);
        result = cmp::max(not_added, added);
    }
    memo[n][capacity as usize] = result;
    return result;
}

fn make_rand_vec(size: i32) -> Vec<i32> {
    let mut rng = rand::thread_rng();
    let mut array: Vec<i32> = Vec::new();
    for _ in 0..size {
        array.push(rng.gen_range(0, size) as i32);
    }
    return array
}
